minimailgun/tasks.py

```diff
@@ -0,0 +1,76 @@
+
+import dns.resolver
+import operator
+import socket
+from celery import Celery, group
+from celery.utils.log import get_task_logger
+from dns.resolver import Timeout, NoAnswer, NXDOMAIN, YXDOMAIN, NoNameservers
+from smtplib import SMTP, SMTPException
+from minimailgun.config import config
+from minimailgun.store import store
+
+
+celery = Celery(__name__, include=['minimailgun.tasks'])
+celery.conf.update({key.upper(): value for key, value in config['celery'].items()})
+log = get_task_logger(__name__)
+
+
+class DNSLookupError(LookupError):
+    pass
+
+
+@celery.task(ignore_result=True)
+def handle_new_message(id):
+    message = store.get_mail_by_id(id)
+    sendmail_jobs = group(sendmail.si(message, value['email']) for recipient, value in message['_recipients'].items())
+    sendmail_jobs()
+
+
+@celery.task(
+    ignore_result=True,
+    bind=True,
+    default_retry_delay=config['mail_properties']['retry_time'],
+    max_retries=config['mail_properties']['max_retry']
+)
+def sendmail(self, message, rcpt):
```

Comment:
This function seems to be doing too much. Querying MX records and sending a mail are two separate tasks. It will make the logic clearer and the code generally more testable.

As a disclaimer, we usually try to adhere to Unix philosophy when writing code. Each function should preferably do just one thing. Without extremes, of course, but in this case clearly two independent things are going on.